src/views/home/components/transaction-form/transaction-form.view.tsx

```diff
@@ -1,140 +1,167 @@
-import { chains } from "src/constants";
-import { useTransactionContext } from "src/contexts/transaction.context";
-import { Chain, Token, TransactionData } from "src/domain";
+import { Chain, TransactionData } from "src/domain";
+import { useEnvContext } from "src/contexts/env.context";
 
-const defaultTransaction: TransactionData = {
-  from: chains[0],
-  to: chains[1],
-  token: tokens[0],
-  amount: BigNumber.from(0),
-};
+interface TransactionFormProps {
+  onSubmit: (transactionData: TransactionData) => void;
+}
 
-const TransactionForm: FC = () => {
+const TransactionForm: FC<TransactionFormProps> = ({ onSubmit }) => {
   const classes = useTransactionFormtStyles();
+  const env = useEnvContext();
   const [list, setList] = useState<List>();
-  const { setTransaction } = useTransactionContext();
   const [isInvalid, setIsInvalid] = useState(true);
-  const [localTransaction, setLocalTransaction] = useState(defaultTransaction);
-  const ChainFromIcon = localTransaction.from.icon;
-  const ChainToIcon = localTransaction.to.icon;
+  const [transactionData, setTransactionData] = useState<TransactionData>();
 
   const onChainFromButtonClick = (from: Chain) => {
-    setLocalTransaction({ ...localTransaction, from });
-    setList(undefined);
-  };
-  const onChainToButtonClick = (to: Chain) => {
-    setLocalTransaction({ ...localTransaction, to });
-    setList(undefined);
-  };
-  const onTokenClick = (token: Token) => {
-    setLocalTransaction({ ...localTransaction, token });
-    setList(undefined);
+    if (env && transactionData) {
+      const to = env.chains.find((chain) => chain.chainId !== from.chainId);
+
+      if (to) {
+        setTransactionData({ ...transactionData, from, to });
+        setList(undefined);
+      }
+    }
   };
 
+  // const onChainToButtonClick = (to: Chain) => {
+  //   if (transactionData) {
+  //     setTransactionData({ ...transactionData, to });
+  //     setList(undefined);
+  //   }
+  // };
+
+  // const onTokenClick = (token: Token) => {
+  //   if (transactionData) {
+  //     setTransactionData({ ...transactionData, token });
+  //     setList(undefined);
+  //   }
+  // };
+
   const onInputChange = ({ amount, isInvalid }: { amount: BigNumber; isInvalid: boolean }) => {
-    setLocalTransaction({ ...localTransaction, amount });
-    setIsInvalid(isInvalid);
+    if (transactionData) {
+      setTransactionData({ ...transactionData, amount });
+      setIsInvalid(isInvalid);
+    }
   };
-  const onSubmit = (e: React.FormEvent) => {
+
+  const onFormSubmit = (e: React.FormEvent) => {
     e.preventDefault();
-    setTransaction(localTransaction);
+    if (transactionData) {
+      onSubmit(transactionData);
+    }
   };
 
+  useEffect(() => {
+    if (env) {
+      setTransactionData({
+        from: env.chains[0],
+        to: env.chains[1],
+        token: env.ethToken,
+        amount: BigNumber.from(0),
+      });
+    }
+  }, [env]);
+
+  if (!env || !transactionData) {
+    return null;
```

Comment:
I'd move this up, so we only check for `env` and `transactionData` once.

Also, I'd show an error if either `env` or `transactionData` are not present.

```diff
diff --git a/src/views/home/components/transaction-form/transaction-form.view.tsx b/src/views/home/components/transaction-form/transaction-form.view.tsx
index 394d1e2..c62e88a 100644
--- a/src/views/home/components/transaction-form/transaction-form.view.tsx
+++ b/src/views/home/components/transaction-form/transaction-form.view.tsx
@@ -13,6 +13,7 @@ import Button from "src/views/shared/button/button.view";
 import AmountInput from "src/views/home/components/amount-input/amount-input.view";
 import { Chain, TransactionData } from "src/domain";
 import { useEnvContext } from "src/contexts/env.context";
+import { useUIContext } from "src/contexts/ui.context";

 interface TransactionFormProps {
   onSubmit: (transactionData: TransactionData) => void;
@@ -21,67 +22,68 @@ interface TransactionFormProps {
 const TransactionForm: FC<TransactionFormProps> = ({ onSubmit }) => {
   const classes = useTransactionFormtStyles();
   const env = useEnvContext();
+  const { openSnackbar } = useUIContext();
   const [list, setList] = useState<List>();
   const [isInvalid, setIsInvalid] = useState(true);
   const [transactionData, setTransactionData] = useState<TransactionData>();

+  useEffect(() => {
+    if (env) {
+      setTransactionData({
+        from: env.chains[0],
+        to: env.chains[1],
+        token: env.ethToken,
+        amount: BigNumber.from(0),
+      });
+    }
+  }, [env]);
+
+  if (!env) {
+    openSnackbar({
+      type: "error-msg",
+      text: "Env is not ready",
+    });
+    return null;
+  }
+
+  if (!transactionData) {
+    openSnackbar({
+      type: "error-msg",
+      text: "Transaction data is not ready",
+    });
+    return null;
+  }
+
   const onChainFromButtonClick = (from: Chain) => {
-    if (env && transactionData) {
     const to = env.chains.find((chain) => chain.chainId !== from.chainId);

     if (to) {
       setTransactionData({ ...transactionData, from, to });
       setList(undefined);
     }
-    }
   };

   // const onChainToButtonClick = (to: Chain) => {
-  //   if (transactionData) {
   //     setTransactionData({ ...transactionData, to });
   //     setList(undefined);
-  //   }
   // };

   // const onTokenClick = (token: Token) => {
-  //   if (transactionData) {
   //     setTransactionData({ ...transactionData, token });
   //     setList(undefined);
-  //   }
   // };

   const onInputChange = ({ amount, isInvalid }: { amount: BigNumber; isInvalid: boolean }) => {
-    if (transactionData) {
     setTransactionData({ ...transactionData, amount });
     setIsInvalid(isInvalid);
-    }
   };

   const onFormSubmit = (e: React.FormEvent) => {
     e.preventDefault();
-    if (transactionData) {
     onSubmit(transactionData);
-    }
   };

-  useEffect(() => {
-    if (env) {
-      setTransactionData({
-        from: env.chains[0],
-        to: env.chains[1],
-        token: env.ethToken,
-        amount: BigNumber.from(0),
-      });
-    }
-  }, [env]);
-
-  if (!env || !transactionData) {
-    return null;
-  }
-
   return (
-    env &&
-    transactionData && (
     <form className={classes.form} onSubmit={onFormSubmit}>
       <Card className={classes.card}>
         <div className={classes.row}>
@@ -161,7 +163,6 @@ const TransactionForm: FC<TransactionFormProps> = ({ onSubmit }) => {
         />
       )}
     </form>
-    )
   );
 };


```

Reply:
The `useEffect` is called asynchronously, so it could be possible that this code gets executed before setting the `transactionData` when it's available:

```ts
+  if (!transactionData) {
+    openSnackbar({
+      type: "error-msg",
+      text: "Transaction data is not ready",
+    });
+    return null;
+  }
```

Reply:
As I've explained to @adrifdez before, I don't like to mix `JSX` code with other things, so for me things should be in order:

1. Constants
2. Event handlers
3. Effects
4. JSX

I had difficult times understanding what's going on when we were returning `null` or the `Redirect` component between event handlers or constants.

As you and @adrifdez proposed the same, if you both agree I'll change it, but it's something that I don't like too much.

Reply:
Nope. Leave it. You just convinced me.